File: ios/App/Chat/MessageList.swift
// ---------------------------------------------------------------------------
// MessageList — the scrolling chat history, mirroring the Android MessageList
// (android/.../chat/MessageList.kt) and the webui MessageList + ChatView
// (gateway/webui/src/components/chat/message-list.tsx, chat-view.tsx).
//
// A ScrollView of MessageBubbles with gapMsg (32pt) between messages.
// Pin-to-bottom autoscroll: while pinned (default, at bottom), the list follows
// new tokens/messages; scrolling up unpins and holds position; re-entering the
// snap zone re-pins. Powered by FollowLatestState + followLatestOnScroll.
//
// iOS 18+: onScrollGeometryChange drives the pin FSM precisely.
// iOS 17:  always-follow fallback (prior behavior) — no geometry API available.
//
// accessibilityIdentifier `chat-message-list` scopes the e2e bubble assertions.
// ---------------------------------------------------------------------------
import SwiftUI
import UIKit
import MobileSdk

/// The scroll dimensions the pin FSM consumes. Projecting the iOS 18
/// `onScrollGeometryChange` to this (rather than the full ScrollGeometry) lets
/// SwiftUI skip frames where none of the three changed.
private struct ScrollProbe: Equatable {
    let top: Double
    let height: Double
    let clientHeight: Double
}

struct MessageList: View {
    let messages: [ChatMessage]
    /// Animation mode for the live (streaming) assistant bubble's avatar.
    /// Committed bubbles stay `.idle` — only the in-flight cycle's mark animates,
    /// mirroring the webui activeCycleMode binding + the Android activeMarkMode.
    var activeMarkMode: MarkMode = .idle
    /// Display name shown in the meta row above user bubbles.
    var userName: String = "You"

    /// Pin-to-bottom FSM state. iOS 18+ only; ignored on iOS 17 (always-follow).
    @State private var follow = FollowLatestState()

    private static let placeholder = "Start a conversation\u{2026}"
    private static let bottomAnchor = "chat-bottom-anchor"

    var body: some View {
        Group {
            if messages.isEmpty {
                emptyState
            } else {
                list
            }
        }
        // Tap anywhere on the conversation dismisses the keyboard (ChatGPT/Claude
        // style); contentShape makes the empty-state whitespace tappable too.
        // simultaneousGesture (not onTapGesture) so a tap on a markdown link still
        // opens the link AND dismisses the keyboard — a container onTapGesture would
        // swallow the link's own tap. Swipe-down dismissal is handled by
        // .scrollDismissesKeyboard on the list.
        .contentShape(Rectangle())
        .simultaneousGesture(TapGesture().onEnded { dismissKeyboard() })
    }

    /// Resign the first responder so the soft keyboard retracts. iOS has no
    /// system "hide keyboard" affordance (Android's IME bar does); we broadcast
    /// resignFirstResponder rather than thread @FocusState out of the Composer.
    private func dismissKeyboard() {
        UIApplication.shared.sendAction(
            #selector(UIResponder.resignFirstResponder), to: nil, from: nil, for: nil
        )
    }

    private var list: some View {
        ScrollViewReader { proxy in
            scrollContent(proxy: proxy)
                .scrollDismissesKeyboard(.interactively)
                .accessibilityIdentifier("chat-message-list")
                .onAppear { scrollToBottom(proxy, animated: false) }
        }
    }

    @ViewBuilder
    private func scrollContent(proxy: ScrollViewProxy) -> some View {
        if #available(iOS 18, *) {
            ios18ScrollView(proxy: proxy)
        } else {
            ios17ScrollView(proxy: proxy)
        }
    }

    /// iOS 18+: geometry-driven pin FSM. Scrolling up unpins; re-entering the
    /// snap zone re-pins. Auto-scroll fires only while pinned.
    @available(iOS 18, *)
    private func ios18ScrollView(proxy: ScrollViewProxy) -> some View {
        ScrollView {
            messageRows()
        }
        // Project to just the dimensions the pin FSM reads, so SwiftUI dedupes
        // frames where these three are unchanged (a bare `{ $0 }` fires the action
        // on every scroll geometry change, churning @State every frame).
        .onScrollGeometryChange(for: ScrollProbe.self, of: { geo in
            ScrollProbe(
                top: geo.contentOffset.y + geo.contentInsets.top,
                height: geo.contentSize.height,
                clientHeight: geo.containerSize.height
            )
        }) { _, probe in
            follow = followLatestOnScroll(
                follow, top: probe.top, height: probe.height, clientHeight: probe.clientHeight
            )
        }
        .onChange(of: messages.count) { _, _ in
            if follow.pinned { scrollToBottom(proxy) }
        }
        .onChange(of: messages.last?.content) { _, _ in
            if follow.pinned { scrollToBottom(proxy) }
        }
    }

    /// iOS 17 fallback: always-follow (original behavior, no geometry API).
    private func ios17ScrollView(proxy: ScrollViewProxy) -> some View {
        ScrollView {
            messageRows()
        }
        .onChange(of: messages.count) { _, _ in scrollToBottom(proxy) }
        .onChange(of: messages.last?.content) { _, _ in scrollToBottom(proxy) }
    }

    private func messageRows() -> some View {
        LazyVStack(alignment: .leading, spacing: Space.gapMsg) {
            // ChatRow is Identifiable; divider ids are day-keyed, message ids are
            // index-keyed. NOTE: do NOT key on message.ts — the streaming bubble's
            // ts is stamped `clock.nowMs()` fresh on every derive, so a ts-based id
            // would churn the bubble's identity each token and reset the typewriter
            // @State (re-revealing from zero every frame). chatRows() preserves
            // index-only identity for .message rows, keeping this invariant safe.
            ForEach(chatRows(messages)) { row in
                switch row {
                case let .divider(label, _): DayDivider(label: label)
                case let .message(m, i):
                    MessageBubble(message: m, index: i, avatarMode: avatarMode(for: m), userName: userName)
                }
            }
            Color.clear
                .frame(height: 1)
                .id(Self.bottomAnchor)
        }
        .padding(Space.lg)
    }

    /// The live (streaming) assistant bubble's avatar animates with the
    /// voice/cognition state; committed bubbles are static (`.idle`).
    private func avatarMode(for message: ChatMessage) -> MarkMode {
        message.streaming && message.role == "assistant" ? activeMarkMode : .idle
    }

    /// Follow-latest: pin the bottom anchor into view on growth or streaming
    /// token changes. Non-animated on first appear so a restored history lands
    /// at the tail without a scroll animation.
    private func scrollToBottom(_ proxy: ScrollViewProxy, animated: Bool = true) {
        if animated {
            withAnimation(.easeOut(duration: Motion.normal)) {
                proxy.scrollTo(Self.bottomAnchor, anchor: .bottom)
            }
        } else {
            proxy.scrollTo(Self.bottomAnchor, anchor: .bottom)
        }
    }

    private var emptyState: some View {
        VStack {
            Text(Self.placeholder)
                .font(.system(size: TypeScale.base))
                .foregroundStyle(DuskColors.ink3)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding(Space.xl)
    }
}

#Preview {
    let now = Int64(Date().timeIntervalSince1970 * 1000)
    // Yesterday messages trigger a day divider before today's messages.
    let yesterday = now - 86_400_000
    MessageList(messages: [
        ChatMessage(ts: yesterday, role: "user", content: "message from yesterday", streaming: false, cutoffKind: nil, cycleId: nil, tools: []),
        ChatMessage(ts: now, role: "user", content: "hello", streaming: false, cutoffKind: nil, cycleId: nil, tools: []),
        ChatMessage(ts: now + 1, role: "assistant", content: "Hi! How can I help today?", streaming: false, cutoffKind: nil, cycleId: nil, tools: []),
    ], userName: "Alice")
    .background(DuskColors.bg)
}

File: android/src/main/kotlin/io/sentient/android/chat/MessageList.kt
// ---------------------------------------------------------------------------
// MessageList — the scrolling chat history, mirroring the webui MessageList +
// ChatView (gateway/webui/src/components/chat/message-list.tsx, chat-view.tsx).
//
// A LazyColumn of MessageBubbles with gapMsg (32dp) between messages. Pin-to-
// bottom follow-latest: while pinned (default), scrolls on growth or streaming-
// token change. User scroll-up unpins; re-entering the bottom zone re-pins.
// Empty state shows the "Start a conversation…" placeholder. The list owns no
// state beyond its scroll position — it reads SdkState.messages, passed down.
//
// testTag `chat-message-list` is on the LazyColumn so the e2e driver scopes its
// bubble assertions to the list.
// ---------------------------------------------------------------------------
package io.sentient.android.chat

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshotFlow
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import io.sentient.android.theme.LocalTokens
import io.sentient.mobilesdk.design.Colors
import io.sentient.mobilesdk.sdk.ChatMessage

private const val PLACEHOLDER = "Start a conversation…"

@Composable
fun MessageList(
    messages: List<ChatMessage>,
    activeMarkMode: MarkMode = MarkMode.IDLE,
    userName: String = "You",
    modifier: Modifier = Modifier,
) {
    if (messages.isEmpty()) {
        EmptyState(modifier = modifier)
        return
    }
    val tokens = LocalTokens.current
    val listState = rememberLazyListState()

    // Compute rows once — shared by the LazyColumn items() and the scroll effect
    // so both reference the same list and rows.lastIndex is the correct tail index.
    val rows = chatRows(messages, nowMs = System.currentTimeMillis())

    // Pin-to-bottom follow-latest: mirrors iOS Task 4.1/4.2 semantics.
    // While pinned (default), the list scrolls to the tail on growth or token
    // change. A real user scroll-up unpins and holds position. Re-entering the
    // bottom zone re-pins automatically.
    val atBottom by remember { derivedStateOf { !listState.canScrollForward } }
    var pinned by remember { mutableStateOf(true) }
    var prevFirst by remember { mutableStateOf(0) }
    var prevOffset by remember { mutableStateOf(0) }

    // Unpin on a real user scroll-up; re-pin when back in the bottom zone.
    LaunchedEffect(listState) {
        snapshotFlow { listState.firstVisibleItemIndex to listState.firstVisibleItemScrollOffset }
            .collect { (idx, off) ->
                val movedUp = idx < prevFirst || (idx == prevFirst && off < prevOffset - 1)
                prevFirst = idx
                prevOffset = off
                if (pinned && movedUp && !atBottom) pinned = false
            }
    }
    LaunchedEffect(atBottom) { if (atBottom) pinned = true }

    // Follow latest while pinned (growth or streaming-token change).
    // Scroll to rows.lastIndex — NOT messages.lastIndex — because the LazyColumn
    // renders rows (Msg + Divider items), so messages.lastIndex is the wrong target.
    LaunchedEffect(rows.size, messages.lastOrNull()?.content) {
        if (pinned && rows.isNotEmpty()) listState.animateScrollToItem(rows.lastIndex)
    }

    LazyColumn(
        modifier = modifier
            .fillMaxSize()
            .testTag("chat-message-list"),
        state = listState,
        contentPadding = PaddingValues(
            horizontal = tokens.space.lg,
            vertical = tokens.space.lg,
        ),
        verticalArrangement = Arrangement.spacedBy(tokens.space.gapMsg),
    ) {
        items(
            rows,
            // Index-only key for Msg rows — the streaming bubble's ts is stamped fresh on
            // every SDK derive, so a ts-based key would churn row identity each token and
            // reset the typewriter rememberTypewriterText @State (re-revealing from zero
            // every frame). History is append-only so the index is stable; the streaming
            // bubble is always the last Msg. Matches the iOS Task 5.1/4.2 fix.
            key = { row ->
                when (row) {
                    is ChatRow.Divider -> "div-${row.key}"
                    is ChatRow.Msg -> "msg-${row.index}"
                }
            },
        ) { row ->
            when (row) {
                is ChatRow.Divider -> DayDivider(row.label)
                is ChatRow.Msg -> {
                    // The live (streaming) assistant bubble's avatar animates with the
                    // voice/cognition state; committed bubbles are static (IDLE). Mirrors
                    // the webui activeCycleMode binding (only the in-flight cycle's mark).
                    val mode = if (row.message.streaming && row.message.role == "assistant") {
                        activeMarkMode
                    } else {
                        MarkMode.IDLE
                    }
                    MessageBubble(
                        message = row.message,
                        index = row.index,
                        avatarMode = mode,
                        userName = userName,
                        modifier = Modifier.fillMaxWidth(),
                    )
                }
            }
        }
    }
}

@Composable
private fun EmptyState(modifier: Modifier = Modifier) {
    val tokens = LocalTokens.current
    Box(
        modifier = modifier
            .fillMaxSize()
            .padding(tokens.space.xl),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = PLACEHOLDER,
            color = Color(Colors.ink3),
            fontSize = tokens.type.base,
        )
    }
}
